dns.go

```diff
@@ -20,29 +20,32 @@ import (
 // OpportunisticResolver opportunistically tries encrypted DNS over TLS
 // using the local resolver.
 var OpportunisticResolver = &net.Resolver{
-	Dial:     opportunisticDial,
+	Dial:     getAnOpportunisticDialerFromDialFunc(nil),
 	PreferGo: true,
 }
 
-func opportunisticDial(ctx context.Context, network, address string) (net.Conn, error) {
-	host, port, _ := net.SplitHostPort(address)
-	if (port == "53" || port == "domain") && notBadServer(address) {
-		deadline, ok := ctx.Deadline()
-		if ok && deadline.After(time.Now().Add(2*time.Second)) {
-			var d net.Dialer
-			d.Timeout = time.Second
-			tlsAddr := net.JoinHostPort(host, "853")
-			tlsConf := tls.Config{InsecureSkipVerify: true}
-			conn, _ := tls.DialWithDialer(&d, "tcp", tlsAddr, &tlsConf)
-			if conn != nil {
-				return conn, nil
+func getAnOpportunisticDialerFromDialFunc(dial DialFunc) DialFunc {
```

Comment:
Do we really need this?

This is only being used internally to resolve the addresses of the DNS server you want to use. Can't you specify those addresses instead?

Reply:
> Can't you specify those addresses instead?

It's what we will do.
I've done that because I'm not sure to deal with the other situation, does Opportunistic Resolving should error if a `DialFunc` is set but no addresses, that for me a weird side effect of `DialFunc` ? (I'm confident about: silentely still trying is an absolute no go, peoples could do un proxied requests without knowing it.)

Reply:
Humm. Not sure I buy the privacy angle here. This is a _single_ request to get your (future) resolver's addresses.

Also, what happens if my system's resolver is `192.168.1.254`? This either breaks (because you can't use Tor to reach `192.168.1.254`, can you?), or it does some how reach `192.168.1.254`, and you'll back at leaking that (_single_) request on the leg that really matters (WAN).

The real fix (in terms of privacy), is to _always_ specify the resolver's addresses, which for DoT you _really_ should.

For DoH, the standard says you shouldn't (which is, IMO, wrong), but this also seems the wrong way to fix it. The "new API" hides (and rightly so) the `Dial` inside an `http.Transport`. So for DoH, I _really don't_ think it makes sense to use that `Dial` for this bit.

The only argument I can see for this, is if the default `net.Dialer`, somehow, doesn't offer connectivity, but yours does (perhaps in something like TinyGo).

I'll still accept it (for DoT) if it works, I'm just unconvinced it is a real privacy gain.